File: answers/leetcode/Reorder List/Reorder List.cpp
// http://fisherlei.blogspot.ca/2013/11/leetcode-reorder-list-solution.html
//@result 13 / 13 test cases passed. Status: Accepted Runtime: 60 ms Submitted: 0 minutes ago You are here! Your runtime beats 91.05% of cpp submissions.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode *head) {
		if (NULL == head) {
			return;
		}
		ListNode *fast = head, *slow = head;
		while (fast) {
			fast = fast->next;
			if (fast) {
				fast = fast->next;
				if (fast) {
					slow = slow->next;
				}
			}
		}
		if (NULL == slow) {
			return;
		}
		ListNode *current = slow;
		ListNode *previous = slow->next;
		current->next = NULL;
		while (previous) {
			ListNode *temp = previous->next;
			previous->next = current;
			current = previous;
			previous = temp;
		}
		ListNode *first = head;
		ListNode *second = current;
		while (first && second && first != second) {
			ListNode *temp = second->next;
			second->next = first->next;
			first->next = second;
			first = second->next;
			second = temp;
		}
    }
};

File: answers/leetcode/Reorder List/Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode left = head;
        ListNode right = head;
        while (right != null && right.next != null && right.next.next != null) {
            left = left.next;
            right = right.next.next;
        }

        ListNode current = left.next;
        left.next = null;
        ListNode dummy = new ListNode(-1);

        while (current != null) {
            ListNode next = current.next;
            current.next = dummy.next;
            dummy.next = current;
            current = next;
        }

        left = head;
        right = dummy.next;
        while (right != null) {
            ListNode next0 = left.next;
            ListNode next1 = right.next;
            left.next = right;
            right.next = next0;
            left = next0;
            right = next1;
        }
    }
}
//Runtime: 1 ms, faster than 99.92% of Java online submissions for Reorder List.
//Memory Usage: 41.6 MB, less than 74.66% of Java online submissions for Reorder List.
